doc/proposals/named-networks.md

```diff
@@ -0,0 +1,187 @@
+# Support for named networks
+
+# Table of contents
+
+- [Introduction](#introduction)
+  - [Goal](#goal-of-this-proposal)
+- [Design](#design)
+  - [Changes in IPAM Config](#changes-in-ipam-config)
+  - [Changes in Modules](#changes-in-modules)
+- [Summary](#summary)
+- [Discussions and Decisions](#discussions-and-decisions)
+
+<hr>
+
+## Introduction
+
+When whereabouts assigns an IP to a Pod this fact is recorded in a CR of kind `IPPool` that has its name derived from the CIDR range in question.
+
+Should the user configure multiple overlapping ranges, it is possible to configure whereabouts to allow assigning duplicate IPs.
+
+However, since the storage of the assignments is done in a CR that is named like the CIDR range, it is not possible to configure *the same CIDR range* twice and have whereabouts assign from the ranges independently.
+
+This is, for example, useful in multi-tenant situations where more than one group is responsible for selecting CIDR ranges.
+
+### Goal of this Proposal
+
+- Allow configuring the same CIDR range multiple times (e.g. in separate multus-`NetworkAttachmentDefinitions`)
+
+<hr>
+
+## Design
+
+The network configuration already has a field `name`:
+
+```json
+{
+      "cniVersion": "0.3.0",
+      "name": "whereaboutsexample",
+      "type": "macvlan",
+      "master": "eth0",
+      "mode": "bridge",
+      "ipam": {
+        "type": "whereabouts",
+        "range": "192.168.2.225/28"
+      }
+}
+```
+
+That is also parsed into the internal representation of the `IPAMConfig`.
+
+This proposal shows three schemes of implementing using that name to distinguish the assignment of IPs:
+
+1. Store the CR with the name of the network configuration instead of the canonicalized CIDR range
+2. Store the CR with the name of the network configuration prepended (or appended) to the canonicalized CIDR range
+3. Add a new field into the `IPAMConfig` to allow users to decide when whereabouts should use the name or the CIDR range for identifying the configuration
+
+### Analysis of the proposed schemes
+
+#### Store the CR under the name of the network configuration
+
+<table>
+<tr>
+<th>Pros</th>
+<th>Cons</th>
+</tr>
+<tr>
+<td>
+:green_circle: Clean design<br/>
+:green_circle: Ranges are easy to find during debugging<br/>
+:green_circle: No more IP-to-string canonicalization<br/>
+</td>
+<td>
+:red_circle: Not backwards compatible, existing installation would need to carefully migrate the existing `IPPool`s to not get duplicate IPs<br/>
+:red_circle: Unclear semantics when two ranges with the same name but different CIDR-ranges are created<br/>
```

Comment:
> how can you have two ranges with the same name in the same namespace ?

In my testing all `IPPools` where always created in the `kube-system`-namespace -- am I misremembering things?

> I am assuming a new pod will only have access to net-attach-defs in its namespace. I would personally even favor a solution where it wouldn't be able to access the cluster's default namespace - unless the pod was created there.

For our use-case we actually need to be able to (also) have central shared `NetworkAttachmentConfiguration`s.

I will re-write the proposal:
* `IPPools` are created in the namespace of the `NetworkAttachmentDefinition`
* Cross-Namespace access is allowed if multus allows it

Reply:
> > how can you have two ranges with the same name in the same namespace ?
> 
> In my testing all `IPPools` where always created in the `kube-system`-namespace -- am I misremembering things?

That's correct, but I don't see how this invalidates my claim: you list as a drawback "Unclear semantics when two ranges with the same name but different CIDR-ranges are created"

Hence the question: how can you have a pool with the same in the same common namespace ? Maybe I'm missing something, but I don't understand your point. I'm assuming this drawback applies to the option where we redesign the pools to have a name instead of a range.

> 
> > I am assuming a new pod will only have access to net-attach-defs in its namespace. I would personally even favor a solution where it wouldn't be able to access the cluster's default namespace - unless the pod was created there.
> 
> For our use-case we actually need to be able to (also) have central shared `NetworkAttachmentConfiguration`s.
> 
> I will re-write the proposal:
> 
> * `IPPools` are created in the namespace of the `NetworkAttachmentDefinition`
> * Cross-Namespace access is allowed if multus allows it

Make sure to add this to a requirement section / something alike.

Reply:
>> In my testing all IPPools where always created in the kube-system-namespace -- am I misremembering things?
>
> That's correct, but I don't see how this invalidates my claim: you list as a drawback "Unclear semantics when two ranges with the same name but different CIDR-ranges are created"

You can create two `NetworkAttachmentDefinition`s called `example` in two namespaces -- whereabouts would then try to create two `IPPools` called `example` -- possibly with different CIDRs defined.

So if we go with the "only the name of the network goes into the IPPool-name", then we should possibly put the IPPool in the same namespace as the `NetworkAttachmentDefinition`?

Reply:
> So if we go with the "only the name of the network goes into the IPPool-name", then we should possibly put the IPPool in the same namespace as the `NetworkAttachmentDefinition`?

I had assumed that, yes.

Is there any particular reason to keep all the pools in the same namespace ? @dougbtv

Reply:
:heavy_check_mark: